palace/linalg/errorestimator.cpp

```diff
@@ -0,0 +1,332 @@
+    smooth_projector(h1_fespaces, iodata.solver.linear.tol, 200, 1,
+                     iodata.solver.pa_order_threshold),
+    smooth_flux(h1_fespaces.GetFinestFESpace().GetTrueVSize()),
+    flux_rhs(h1_fespaces.GetFinestFESpace().GetTrueVSize()),
+    field_gf(&h1_fespaces.GetFinestFESpace()),
+    smooth_flux_gf(&h1_fespaces.GetFinestFESpace())
+{
+}
+
+ErrorIndicators GradFluxErrorEstimator::ComputeIndicators(const Vector &v) const
+{
+  auto &h1_fespace = h1_fespaces.GetFinestFESpace();
+  const int sdim = h1_fespace.GetMesh()->SpaceDimension();
+  field_gf.SetFromTrueDofs(v);
+  const int nelem = h1_fespace.GetNE();
+
+  Vector estimates(nelem);
+  estimates = 0.0;
+  double normalization = 0.0;
+
+  // Coefficient for computing the discontinuous flux., i.e. (Vᵢ, (ϵ ∇ ϕ)ᵢ).
+  GradFluxCoefficient coef(field_gf, mat_op);
+  for (int c = 0; c < sdim; c++)
+  {
+    coef.SetComponent(c);
+    {
+      // Given the RHS vector of non-smooth flux component, compute fᵢ = M⁻¹ f̂ᵢ.
+      Mpi::Print("Computing smooth flux approximation of component {}\n", c);
+      BlockTimer bt0(Timer::ESTSOLVE);
+      mfem::LinearForm rhs(&h1_fespace);
+      rhs.AddDomainIntegrator(new mfem::DomainLFIntegrator(coef));
+      rhs.UseFastAssembly(false);
+      rhs.Assemble();
+      h1_fespace.GetProlongationMatrix()->MultTranspose(rhs, flux_rhs);
+      smooth_projector.Mult(flux_rhs, smooth_flux);
+    }
+    smooth_flux_gf.SetFromTrueDofs(smooth_flux);
+    smooth_flux_gf.ExchangeFaceNbrData();
+    for (int e = 0; e < h1_fespace.GetNE(); e++)
+    {
+      auto &T = *h1_fespace.GetElementTransformation(e);
+      // integration order 2p + q
+      const auto &ir = mfem::IntRules.Get(T.GetGeometryType(),
+                                          2 * h1_fespace.GetFE(e)->GetOrder() + T.Order());
+      for (const auto &ip : ir)
+      {
+        T.SetIntPoint(&ip);
+        double smooth_val = smooth_flux_gf.GetValue(e, ip);
+        double coarse_val = coef.Eval(T, ip);
+        const double w_i = ip.weight * T.Weight();
+        estimates[e] += w_i * std::pow(smooth_val - coarse_val, 2.0);
+        normalization += w_i * std::pow(coarse_val, 2.0);
+      }
+    }
+    if constexpr (false)
+    {
+      // Debugging branch generates some intermediate fields for paraview.
+      mfem::ParaViewDataCollection paraview("debug_coeff" + std::to_string(c),
+                                            h1_fespace.GetParMesh());
+      paraview.RegisterCoeffField("Flux", &coef);
+      paraview.RegisterField("SmoothFlux", &smooth_flux_gf);
+
+      mfem::L2_FECollection est_fec(0, sdim);
+      mfem::ParFiniteElementSpace est_fespace(h1_fespace.GetParMesh(), &est_fec);
+      mfem::ParGridFunction est_field(&est_fespace);
+      est_field.SetFromTrueDofs(estimates);
+
+      paraview.RegisterField("ErrorIndicator", &est_field);
+      paraview.Save();
+    }
+  }
+  for (auto &e : estimates)
+  {
+    e = std::sqrt(e);
```

Comment:
Can move into the loop above?

Reply:
Each element is visited twice, once for real and imaginary, accumulating the square, so this has to be done afterwards.

Reply:
Introduce `CwiseSqrt`
https://github.com/awslabs/palace/pull/85/commits/eca57eb96d4216cf5d5ac1ed3cc1c9532f069225